### PR TITLE
Add react-testing-library to react variant

### PR DESCRIPTION
**Pull request by mischa-s**
This pull request adds [react-testing-library](https://testing-library.com/docs/react-testing-library/intro/) to the react variant.

It includes:

- an example test for UI interactions
- documentation in `readme` for devs who are just getting started with react-testing-library
- eslint plugins for react-testing-library
- package.json scripts for single test run and watching the test suite

### CODE REVIEW

File: .eslintrc.js
Comment (G-Rath) [inline .eslintrc.js:10]:
```suggestion
    browser: true
  },
```

`eslint-plugin-jest` enables this.

File: .eslintrc.js
Comment (G-Rath) [inline .eslintrc.js:15]: We should use our `ackama/jest` config, which should be applied to only test files using `overrides` - this also applies to `testing-library` and co configs + plugins.

i.e

```
  ...
  extends: ['ackama', 'ackama/@typescript-eslint'],
  ignorePatterns: [],
  overrides: [
    {
      files: ['test/**'],
      extends: ['ackama/jest'],
      rules: { 'jest/prefer-expect-assertions': 'off' }
    }
  ],
  ...
```

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [inline variants/frontend-base/js-lint/template.rb:5]: 🐘 I think this should be refactored to use an `w%` array, like I did with the [typescript varient](https://github.com/ackama/rails-template/pull/207/files#diff-23eb3fd5a078907cdb393b92e280dc3c885f44e38857baf6b6f6b81e63535576R6-R22).

I'm happy for this to be done in its own PR.

File: variants/frontend-react/app/frontend/components/HelloWorld.jsx
Comment (G-Rath) [inline variants/frontend-react/app/frontend/components/HelloWorld.jsx:7]: Not the biggest fan of this as a new function will be created every time the component is rendered; would prefer using `useCallback` instead - while overkill in this component, I'd like to encourage its usage for those less familiar with React & hooks.

Also would prefer using arrow functions rather than a hoisted function, to make it easier to reason mentally (since right now its being hoisted to the top of the component, before `setGreeting` is declared).

Finally, would prefer calling the parameter `event` rather than just `e` to be clearer what the expected argument is.

File: variants/frontend-react/template.rb
Comment (G-Rath) [inline variants/frontend-react/template.rb:60]:
```suggestion
    <%= react_component("HelloWorld", { initialGreeting: "Hello from react-rails." }) %>
  ERB
```

🐘

File: variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx
Comment (G-Rath) [inline variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx:20]:
```suggestion
    await findByText(/hello ackama/iu);

    userEvent.type(
      screen.getByRole('textbox', {
        name: /change the greeting/iu
      }),
      'Hello from the other side'
    );

    expect(screen.getByText(/hello from the other side/iu)).toBeInTheDocument();
```

🐘 super nity - I find having newlines around expect blocks and keywords like `await` helps improve readablity

File: variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx
Comment (G-Rath) [inline variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx:7]: nice! ❤️ 

This is a great test example - much better than what I came up with last time I played with testing library 😅

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [inline variants/frontend-base/js-lint/template.rb:33]:
```suggestion
  "watch-tests" => "jest --watch"
}
```

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [inline variants/frontend-base/js-lint/template.rb:17]: I'd prefer to use `jest.config.js`, as I think it makes it easier to find (especially when you're not as familiar with the stack, since its not as obvious to check `package.json` for config).

This'll also make things more consistent with other projects that'll have more settings, and allows us to use type-checking to assist.

File: README.md
Comment (G-Rath) [inline README.md:236]:
```suggestion
An example react test using [react-testing-library](https://testing-library.com/docs/react-testing-library/intro/) is provided. Before you start adding more tests, it is recommended you read [common-mistakes-with-react-testing-library](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library)
```

🐘

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [reply variants/frontend-base/js-lint/template.rb:5]:
```suggestion
run "yarn add --dev eslint eslint-config-ackama eslint-plugin-node eslint-plugin-import eslint-plugin-prettier prettier prettier-config-ackama eslint-plugin-eslint-comments eslint-plugin-react eslint-plugin-react-hooks eslint-plugin-jsx-a11y jest eslint-plugin-testing-library eslint-plugin-jest-dom eslint-plugin-jest eslint-plugin-jest-formatting"
```

(this'll be required when you add `ackama/jest`)

**Review by G-Rath — CHANGES_REQUESTED** [review]
Got a few things I'd like addressed, but overall looking good!

I'm neutral on where you've actually put the tests - in javascript projects we tend to put them in `test/`, and I think having them at the root would make them more discoverable, *but* this is a whole new thing for us and it'd potentially have to share that space with other testing tools, so on the whole I'm happy with seeing how this goes.

Do you want to have a crack at the typescript version in a followup PR? otherwise I can do it :)
(I'll merge the TS variant now)

File: variants/frontend-react/app/frontend/components/HelloWorld.jsx
Comment (mischa-s) [reply variants/frontend-react/app/frontend/components/HelloWorld.jsx:7]: `useCallback`: This is an interesting one which I think we've discussed previously. While you're 100% correct from a performance perspective, as you say this level of optimisation is not required for components which are this simple. If we follow that to it's conclusion, we end up with the advice "always memoize your functions" which has been debated with the react community but from everything I've seen, the consensus so far is, "only memoize when the gains are meaningful".

Comment (G-Rath) [reply variants/frontend-react/app/frontend/components/HelloWorld.jsx:7]: That's fine - I'm less concerned about if we use `useCallback` here or not vs using a `const` which I feel makes it a lot more obvious that the functions ref lifecycle is tied to that of the components, by way of the variable being local.

(iirc we discussed this previously too)

File: .eslintrc.js
Comment (G-Rath) [inline .eslintrc.js:9]:
```suggestion
  env: { commonjs: true, node: true, browser: true },
```

🐘 might as well restore the original layout to reduce the diff noise

File: .eslintrc.js
Comment (mischa-s) [reply .eslintrc.js:14]: > Got a few things I'd like addressed, but overall looking good!
> 
> I'm neutral on where you've actually put the tests - in javascript projects we tend to put them in `test/`, and I think having them at the root would make them more discoverable, _but_ this is a whole new thing for us and it'd potentially have to share that space with other testing tools, so on the whole I'm happy with seeing how this goes.
> 
> Do you want to have a crack at the typescript version in a followup PR? otherwise I can do it :)
> (I'll merge the TS variant now)

I can see about adding the TS variant tomorrow morning if I get some time

**Comment by mischa-s** [comment]
> Got a few things I'd like addressed, but overall looking good!
> 
> I'm neutral on where you've actually put the tests - in javascript projects we tend to put them in `test/`, and I think having them at the root would make them more discoverable, _but_ this is a whole new thing for us and it'd potentially have to share that space with other testing tools, so on the whole I'm happy with seeing how this goes.
> 
> Do you want to have a crack at the typescript version in a followup PR? otherwise I can do it :)
> (I'll merge the TS variant now)

I can see about adding the TS variant tomorrow morning if I get some time

File: .eslintrc.js
Comment (G-Rath) [reply .eslintrc.js:10]: No rush - it'll probably be worth landing in its own PR to keep the noise down anyway, once this one is merged.

**Comment by G-Rath** [comment]
No rush - it'll probably be worth landing in its own PR to keep the noise down anyway, once this one is merged :)

File: variants/frontend-react/template.rb
Comment (G-Rath) [inline variants/frontend-react/template.rb:11]: 🐘 nity thing that could be worth doing in the future: having this in `test/template.rb`, to match how we've got `lint/template.rb`

File: variants/frontend-react/app/frontend/tests/setupExpectEachTestHasAssertions.js
Comment (G-Rath) [inline variants/frontend-react/app/frontend/tests/setupExpectEachTestHasAssertions.js:5]: This should have a trailing newline.

**Review by G-Rath — CHANGES_REQUESTED** [review]
Looking good!

2 things:

1. This is causing the TS variant to fail:

```
/home/runner/work/rails-template/rails-template/template-test/dummy/react-typescript-test-app/app/frontend/tests/HelloWorld.spec.jsx
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: app/frontend/tests/HelloWorld.spec.jsx.
The file must be included in at least one of the projects provided
```

That's a pretty standard error, which'll be because the `.spec.jsx` file isn't being converted to TypeScript.

I'd really like the TS support to be landed in its own PR, so for now your best bet is probably to remove the new js files in the typescript variant (don't worry about un-adding the dependencies).

2. I just realised we're applying this to the `js-template`, when it should be being applied to the `react` template, since its only for when we're generating react 🤦 

That's actually a bit problematic as the `js-lint` template is applied after the react one.
I think the best bet is probably to modify `js-lint` to test if `jest.config.js` exists, and if so add the extra `jest` based config.

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [inline variants/frontend-base/js-lint/template.rb:5]: I just realised that the `js-lint` template is applied regardless of if we're using `react` or not, which we shouldn't do.

I think the best thing to do would be to have an existence check for the `jest.config`, and then if that's there this template should do a bit of extra work.

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [inline variants/frontend-base/js-lint/template.rb:25]: ditto for these lines regarding the `react` template variant.

Since this is a hash, you should be able to easily solve this:

```
package_json["scripts"] = {
  "js-lint" => "eslint . --ignore-pattern '!.eslintrc.js' --ext js,ts,tsx,jsx",
  "js-lint-fix" => "eslint . --ignore-pattern '!.eslintrc.js' --ext js,ts,tsx,jsx --fix",
  "format-check" => "prettier --check './**/*.{css,scss,json,md,js,ts,tsx,jsx}'",
  "format-fix" => "prettier --write './**/*.{css,scss,json,md,js,ts,tsx,jsx}'",
  "scss-lint" => "stylelint '**/*.{css,scss}'",
  "scss-lint-fix" => "stylelint '**/*.{css,scss}' --fix"
  "scss-lint-fix" => "stylelint '**/*.{css,scss}' --fix"
}

package_json.merge!({
  "test" => "jest /app/frontend",
  "watch-tests" => "jest --watch"
}) if is_using_jest
```

File: variants/frontend-base/js-lint/template.rb
Comment (mischa-s) [reply variants/frontend-base/js-lint/template.rb:5]: are you suggesting not runing the `js-lint` template at all if we're not using react - or just wrapping this new stuff in a conditional?

Comment (G-Rath) [reply variants/frontend-base/js-lint/template.rb:5]: ofph that's a good pont 🤦 

Ideally we probably should only be configuring react in the linting config if the react variant is used, but since we're not already, I think you can ignore this comment :D

(that's probably something to implement as a follow-up)

File: variants/frontend-base/js-lint/template.rb
Comment (G-Rath) [reply variants/frontend-base/js-lint/template.rb:25]: this can be ignored per [this comment](https://github.com/ackama/rails-template/pull/216#discussion_r579915917)

File: variants/frontend-react/jest.config.js
Comment (G-Rath) [inline variants/frontend-react/jest.config.js:9]:
```suggestion
  setupFilesAfterEnv: [
    '@testing-library/jest-dom/extend-expect',
    'app/frontend/tests/setupExpectEachTestHasAssertions.js'
  ]
```

This'll mean we don't have to have this import in every test.

(Technically we don't have to have the `/extend-expect` at the end, but think it makes it more obvious what this is doing)

File: variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx
Comment (G-Rath) [inline variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx:4]: This can be [dropped](https://github.com/ackama/rails-template/pull/216/files#r583919507)

File: variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx
Comment (G-Rath) [inline variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx:8]: Currently this test is actually testing two things that could be in their own separate tests - I think it would be good to split it up so that it's more focused, as usually that makes for better testing.

I'd generally structure the tests like so:

* `describe 'HelloWorld'`
  * `it 'renders the initial greeting'` (tests that `initialGreeting` is rendered)
  * `describe 'when the greeting is changed'`
    * `it 'renders the new greeting'` (tests that the greeting changes)
  * `describe 'when the greeting is cleared'`
    * `it 'falls back to the initial greeting'` (tests that blanking the input has it fall back to the `initialGreeting`)

I've chucked in that last one as a possible third test that'd be easy to implement into the component to give a good reason for a third test - feel free not to implement it; some good alternatives could be either having the `input` value default to `initialGreeting` (so showcasing how to test that an input field has a default value based off a prop), or adding validation to `input` so that it's not allowed to be blank (meaning you'd showcase how you test a more complex thing like validation errors).

Comment (G-Rath) [reply variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx:8]: > I think `findByText` is actually being used as an assertion, which is fine

In saying that, I think it would be good to try and stick to the `expect` matchers as much as possible, which seems to be possible:

```
describe('Stats', () => {
  describe('when the api request is in progress', () => {
    it('shows a loading state', () => {
      fetchPullRequestStatsMock.mockReturnValue(new Promise(() => undefined));

      const { container } = render(<Stats />);

      expect(container).toHaveTextContent(/Loading/iu);
    });
  });

  describe('when the api request resolves', () => {
    it('shows the stats', async () => {
      // useAsync currently requires the Promise to actually be an instanceof the Promise class
      fetchPullRequestStatsMock.mockImplementation(async () =>
        Promise.resolve({
          numberOfPullRequestsClosed: 3,
          averageTimePullRequestsSpendOpen: 1234
        })
      );

      const { container, queryByText } = render(<Stats />);

      await waitForElementToBeRemoved(() => queryByText(/Loading/iu));

      expect(container).not.toHaveTextContent(/Loading/iu);

      expect(container).toHaveTextContent(/PRs Closed: 3/iu);
      expect(container).toHaveTextContent(/Average Time Spent Open: 1234/iu);
    });
  });
})
```

This'd mean we wouldn't have to loosen our linting rules as much, and let us use `expect.hasAssertions()` for every test.

What do you think?

**Comment by G-Rath** [comment]
This is currently blocked by testing-library not completely supporting Jest v27, due to changes we made in `setTimeout` & co. Will get back to it "soon".

**Comment by mischa-s** [comment]
closing in favour of https://github.com/ackama/rails-template/pull/277

**Pull request closed without merging**
